jupyter_core/command.py

```diff
@@ -72,8 +73,17 @@ def list_subcommands():
             continue
         for name in names:
             if name.startswith('jupyter-'):
-                subcommands.add(name.split('-')[1])
-    return subcommands
+                subcommand_tuples.add(tuple(name.split('-')[1:]))
+    # build a set of subcommand strings, excluding subcommands whose parents are defined
+    subcommands = set()
+    for sub_tup in subcommand_tuples:
+        for i in range(1, len(sub_tup)):
+            # Only include `jupyter-foo-bar` if `jupyter-foo` is not already present
+            if sub_tup[:i] in subcommand_tuples:
+                break
+        else:
```

Comment:
I think this check might be clearer with an any():

`if not any(sub_tup[:i] in subcommand_tuples for i in range(1, len(sub_tup))):`